### PR TITLE
Fix deletion using RegionParticleIterators for VerletClusterCells

### LINKED ISSUE #435: Deleting particles when using a RegionParticleIterator breaks VerletClusterCells
**Is your feature request related to a problem? Please describe.**
For the two cluster containers, deleting a particle using RegionParticleIterators is especially problematic. Particles are usually deleted using a swap to the end of a cell. For the Cluster containers the assumption of sorted particles, that is used in the RegionParticleIterators, is thus broken and errors are likely to occur.

**Describe the solution you'd like**
Adding a test for that!

**Additional context**
Probably VerletClusterCells is broken for this! As it implements exactly such an optimization.

### PR DESCRIPTION
**Pull request by SteffenSeckler**
# Description

Implements the test requested in #435.
As expected, this test fails for VerletClusterCells.

## Related Pull Requests

- Depends on #430 

## Resolved Issues

- [x] fixes #435 

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes.

- [x] New test for deletion of RegionParticleIterators

### CODE REVIEW

File: src/autopas/containers/verletClusterCells/VerletClusterCellsParticleIterator.h [hunk 1]
```diff
@@ -261,17 +275,26 @@ class VerletClusterCellsRegionParticleIterator
...
+          this->_iteratorWithinOneCell = std::lower_bound(
+              (*this->_vectorOfCells)[this->_cellId]._particles.begin(),
+              (*this->_vectorOfCells)[this->_cellId]._particles.begin() + this->getDummyStartbyIndex(this->_cellId),
+              _startRegion[2] - _skin, [](const Particle &a, const double b) { return a.getR()[2] < b; });
+          this->_cellEnd = std::upper_bound(
+              (*this->_vectorOfCells)[this->_cellId]._particles.begin(),
+              (*this->_vectorOfCells)[this->_cellId]._particles.begin() + this->getDummyStartbyIndex(this->_cellId),
+              _endRegion[2] + _skin, [](const double b, const Particle &a) { return b < a.getR()[2]; });
+        } else {
+          // If the iterator is modifiable, then we can delete particles, which breaks the sorted assumption and thus
```
Comment (FG-TUM) [inline src/autopas/containers/verletClusterCells/VerletClusterCellsParticleIterator.h:293]:
```suggestion
          // If the iterator is modifiable, we can delete particles, which breaks the sorted assumption and thus
```

File: tests/testAutopas/tests/autopasInterface/IteratorTest.cpp [hunk 2]
```diff
@@ -453,6 +453,84 @@ TEST_P(IteratorTest, testOpenMPRegionIteratorsHaloOnly) {
   }
 }
 
+/**
+ * The main idea of this test is to check whether deletion through a RegionIterator doesn't do stupid things.
```
Comment (FG-TUM) [inline tests/testAutopas/tests/autopasInterface/IteratorTest.cpp:457]: The idea of a test description is to describe the test.....

File: tests/testAutopas/tests/autopasInterface/IteratorTest.cpp [hunk 3]
```diff
@@ -453,6 +453,84 @@ TEST_P(IteratorTest, testOpenMPRegionIteratorsHaloOnly) {
...
+  std::conditional_t<testConstIterators, const autopas::AutoPas<Molecule, FMCell> &,
+                     autopas::AutoPas<Molecule, FMCell> &>
+      autoPasRef = autoPas;
+
+  autoPas.setAllowedContainers(std::set<autopas::ContainerOption>{containerOption});
+  autoPas.setAllowedCellSizeFactors(autopas::NumberSetFinite<double>(std::set<double>({cellSizeFactor})));
+
+  defaultInit(autoPas);
+
+  size_t numParticles = 200;
```
Comment (FG-TUM) [inline tests/testAutopas/tests/autopasInterface/IteratorTest.cpp:473]: do we really need to test 200 particles here?

File: tests/testAutopas/tests/autopasInterface/IteratorTest.cpp [hunk 4]
```diff
@@ -453,6 +453,84 @@ TEST_P(IteratorTest, testOpenMPRegionIteratorsHaloOnly) {
...
+  EXPECT_EQ(afterParticles.size(), shouldBeParticles.size());
+}
+
+TEST_P(IteratorTest, RegionIteratorDeletion) {
+  auto [containerOption, cellSizeFactor, testConstIterators, priorForceCalc] = GetParam();
+  if (testConstIterators) {
+    testRegionIteratorDeletion<true>(containerOption, cellSizeFactor, priorForceCalc);
+  } else {
+    testRegionIteratorDeletion<false>(containerOption, cellSizeFactor, priorForceCalc);
+  }
```
Comment (FG-TUM) [inline tests/testAutopas/tests/autopasInterface/IteratorTest.cpp:531]: I would prefer there to be two TEST_P, one for each of those cases. That way we see more directly what went wrong.

File: tests/testAutopas/tests/autopasInterface/IteratorTest.cpp [hunk 5]
```diff
@@ -453,6 +453,84 @@ TEST_P(IteratorTest, testOpenMPRegionIteratorsHaloOnly) {
...
+    if (deletedParticles.find(id) == deletedParticles.end()) {
+      // Has not been deleted!
+      shouldBeParticles.insert(id);
+    }
+  }
+
+  // check whether they are still there.
+  std::vector<unsigned long> afterParticles;
+  for (auto &particle : autoPasRef) {
+    EXPECT_NE(shouldBeParticles.find(particle.getID()), shouldBeParticles.end()) << "id:" << particle.getID();
```
Comment (FG-TUM) [inline tests/testAutopas/tests/autopasInterface/IteratorTest.cpp:518]: why not print the whole particle

File: tests/testAutopas/tests/autopasInterface/IteratorTest.cpp (hunk 4 above)
Comment (SteffenSeckler) [reply tests/testAutopas/tests/autopasInterface/IteratorTest.cpp:531]: invalid

File: tests/testAutopas/tests/autopasInterface/IteratorTest.cpp (hunk 3 above)
Comment (SteffenSeckler) [reply tests/testAutopas/tests/autopasInterface/IteratorTest.cpp:473]: just for safety :) s.t. we really have multiple particles that are deleted and also particles after the deleted particles in the same towers/cells.

File: tests/testAutopas/tests/autopasInterface/IteratorTest.cpp (hunk 5 above)
Comment (SteffenSeckler) [reply tests/testAutopas/tests/autopasInterface/IteratorTest.cpp:518]: we anyways only safe the particle id.

**Pull request merged**